google/cloud/pubsub/internal/message_carrier.cc

```diff
@@ -0,0 +1,62 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include "google/cloud/pubsub/internal/message_carrier.h"
+#include "google/cloud/pubsub/message.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/noexcept_action.h"
+#include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/options.h"
+#include "absl/strings/match.h"
+#include <opentelemetry/context/propagation/global_propagator.h>
+#include <opentelemetry/context/propagation/text_map_propagator.h>
+#include <opentelemetry/trace/context.h>
+#include <opentelemetry/trace/propagation/http_trace_context.h>
+#include <opentelemetry/trace/provider.h>
+#include <opentelemetry/trace/semantic_conventions.h>
+#include <opentelemetry/trace/span_metadata.h>
+#include <opentelemetry/trace/span_startoptions.h>
+namespace google {
+namespace cloud {
+namespace pubsub_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+opentelemetry::nostd::string_view MessageCarrier::Get(
+    opentelemetry::nostd::string_view key) const noexcept {
+  auto message_attributes = message_.attributes();
+
+  auto it = message_attributes.find(
+      absl::StrCat("googclient_", std::string(key.data(), key.size())));
+  if (it != message_attributes.end()) {
+    return opentelemetry::nostd::string_view(it->second);
+  }
+  return opentelemetry::nostd::string_view{};
```

Comment:
I'm not sure how to use `internal::NoExceptAction` with a return value. If I wrap this in `internal::NoExceptAction` it returns a bool, not the value from the lambda

Reply:
There is another overload that might be relevant to you: https://github.com/googleapis/google-cloud-cpp/blob/31722a166574ec86af24cdfdc23a8a359638e3c5/google/cloud/internal/noexcept_action.h#L43

example usage: https://github.com/googleapis/google-cloud-cpp/blob/64fe5632c2dc6ceb16d628be52d6f120f86d343b/google/cloud/opentelemetry/trace_exporter.cc#L46-L51